design/automatic-network-configuration.md

```diff
@@ -0,0 +1,801 @@
+# Automatic network configuration
+
+## Status
+
+provisional
+
+## Summary
+
+This proposal is to expand the scope of Metal³ to include an API to
+manage physical network devices.
+
+## Motivation
+
+Metal³ follows the paradigm of Kubernetes Native Infrastructure (KNI),
+which is an approach to use Kubernetes to manage underlying infrastructure.
+Managing the configuration of some physical network devices is closely
+related to managing physical hosts. As bare metal hosts are provisioned
+or later repurposed, there may be corresponding physical network changes
+that must be made, such as reconfiguring a ToR switch port. If the provisioning
+of the physical host is managed through a Kubernetes API, it would be
+convenient to be able to reconfigure related network devices using a
+similar API.
+
+## Goals
+
+- Define a Kubernetes API for configuring network device.
+- Automatically configure the network infrastructure for a host when
+adding it to a cluster.
+- Do network configuration when deprovisioning the hosts.
+- Design a network abstraction that can represent any of the target networking
+configuration, independently of the controller used.
+
+## Non-Goals
+
+- Implement a network controller which is aimed at being an integration with
+an existing controller only.
+- Implement a solution for a specific underlying infrastructure.
+- Implement a solution for link aggregation in the future.
+- Implement a solution for lldp in the future.
+
+## Proposal
+
+This document proposes to add a new mechanic to automatically perform physical
+network device configuration before provisioning a BareMetalHost. First we
+added the NetworkConfiguration operator to implement the network configuration
+of the device. Then we added some methods to CAPM3 to implement together. When
+network configuration is required, CAPM3 will process some content through
+additional methods, and then trigger the NetworkConfiguration operator to
+perform network configuration.
+
+### User Stories
+
+#### Story 1
+
+As a consumer of Metal³, when adding a machine resource to the cluster, according
+to the network configuration of the cluster, the physical host is connected to the
+corresponding logical layer-2 network segment.
```

Comment:
User stories are usually formulated in terms of users' desires ("I want my traffic isolated from ..."), not technical implementation.